onnxruntime/core/providers/webnn/builders/impl/attention_op_builder.cc

```diff
@@ -0,0 +1,568 @@
+  emscripten::val attn_mask = model_builder.GetBuilder().call<emscripten::val>("where", condition, value_one_constant,
+                                                                               finfo_min_constant, common_options);
+
+  common_options.set("label", node.Name() + "/GQA/attn_mask/softmax_input");
+  emscripten::val softmax_input =
+      model_builder.GetBuilder().call<emscripten::val>("add", div_output, attn_mask, common_options);
+
+  common_options.set("label", node.Name() + "/GQA/attn_mask/softmax_input");
+  int32_t softmax_axis = 3;
+  emscripten::val softmax_output =
+      model_builder.GetBuilder().call<emscripten::val>("softmax", softmax_input, softmax_axis, common_options);
+
+  common_options.set("label", node.Name() + "/GQA/qkv/matmul_2");
+  emscripten::val attn_output =
+      model_builder.GetBuilder().call<emscripten::val>("matmul", softmax_output, true_present_value, common_options);
+
+  options.set("permutation", emscripten::val::array(std::vector<uint32_t>({0, 2, 1, 3})));
+  options.set("label", node.Name() + "/GQA/qkv/transpose");
+  emscripten::val transposed_attn_output =
+      model_builder.GetBuilder().call<emscripten::val>("transpose", attn_output, options);
+
+  common_options.set("label", node.Name() + "/GQA/qkv/reshape");
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>(
+      "reshape", transposed_attn_output, emscripten::val::array(reshape_output_shape), common_options);
+
+  if (node.OutputDefs()[0]->Type() == onnx::Utils::DataTypeUtils::ToType("float16")) {
+    common_options.set("label", node.Name() + "/GQA/postprocess/cast/output");
+    output =
+        model_builder.GetBuilder().call<emscripten::val>("cast", output, emscripten::val("float16"), common_options);
+  }
+  if (node.OutputDefs()[1]->Type() == onnx::Utils::DataTypeUtils::ToType("float16")) {
+    common_options.set("label", node.Name() + "/GQA/postprocess/cast/present_key");
+    present_key = model_builder.GetBuilder().call<emscripten::val>("cast", present_key, emscripten::val("float16"),
+                                                                   common_options);
+  }
+  if (node.OutputDefs()[2]->Type() == onnx::Utils::DataTypeUtils::ToType("float16")) {
+    common_options.set("label", node.Name() + "/GQA/postprocess/cast/present_value");
+    present_value = model_builder.GetBuilder().call<emscripten::val>("cast", present_value, emscripten::val("float16"),
+                                                                     common_options);
+  }
+
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(output));
+  model_builder.AddOperand(node.OutputDefs()[1]->Name(), std::move(present_key));
+  model_builder.AddOperand(node.OutputDefs()[2]->Name(), std::move(present_value));
+
+  return Status::OK();
+}
+
+// Operator support related.
+
+bool GroupQueryAttentionOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
```

Comment:
Missing check for `total_sequence_length`, which should be an initialization, and its value equals to `max_sequence_length`.